### PR TITLE
Implement explain backend for dynamo

### LINKED ISSUE #102053: [torch.compile] Add `explain` as a backend
[`torch._dynamo.explain`](https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/_dynamo/eval_frame.py#L562) is a useful API for PT2 users. Users can find the number of graphs and graph breaks using this API, giving them actionable messages. One of the pain points is - its a different API than torch.compile. @Chillee  proposed to add it as a backend, simplifying the UX issue. 

~~~
import torch
import torch._dynamo

def fn3(x):
    x = torch.sin(x)
    torch._dynamo.graph_break()
    x = torch.sin(x)
    return x



def fn2(x):
    x = torch.cos(x)
    x = fn3(x)
    x = torch.cos(x)
    return x


def fn1(x):
    x = torch.tan(x)
    torch._dynamo.graph_break()
    x = fn2(x)
    x = torch.tan(x)
    return x

def fn(x):
    x = torch.sigmoid(x)
    x = fn1(x)
    x = torch.sigmoid(x)
    return x


# outs = torch._dynamo.explain(fn, torch.randn(4))
explanation, out_guards, graphs, ops_per_graph, break_reasons, explanation_verbose = torch._dynamo.explain(fn, torch.randn(4))
print(explanation_verbose)
for graph in graphs:
    print(graph.graph.print_tabular())

~~~    

### Short term task

Introduce a new backend - `explain` such that `torch.compile(fn, backend="explain")` gives morally same output as torch._dynamo.explain. We can add the `explain` backend in this file - [torch/_dynamo/backends/debugging.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/backends/debugging.py)

As a start, we can focus on graph breaks and explanation string. We would want the backend to be configurable so that we can expand to other stuff.

### Improving the output

* More details around graph breaks. For example, a graph break deep in the call stack can lead to many graph breaks in the model. The output should explain this part well, and potentially rank the graph breaks.
* Recompiles - And reason of recompile

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by alimoezzi**
Fixes #102053

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

### CODE REVIEW

**Comment by msaroufim** [comment]
Thanks @alimoezzi! The PR looks good lemme review now

To get CI to pass please rebase to main since the `torch.compiler` PR was reverted and install lintrunner

File: torch/_dynamo/backends/debugging.py [hunk 1]
```diff
@@ -18,6 +20,58 @@ def eager(gm, fake_tensor_inputs):
...
+            tabulate(
+                node_specs, headers=["opcode", "name", "target", "args", "kwargs"]
+            ),
+        )
+    if gm.compile_subgraph_reason.graph_break:
+        formatted_stack = "".join(
+            traceback.format_list(gm.compile_subgraph_reason.user_stack)
+        )
+        msg = f"{gm.compile_subgraph_reason.reason}\n{formatted_stack}"
+        print("Graph Break\nBreak reason:\n", msg)
```
Comment (msaroufim) [inline torch/_dynamo/backends/debugging.py:70]: Nit but would be prettier to have something like `Break Reason {i}:`

**Review by msaroufim — APPROVED** [review]
LGTM - cc @voznesenskym for final stamp

**Review by msaroufim — APPROVED** [review]
LGTM - cc @voznesenskym for final stamp

File: torch/_dynamo/backends/debugging.py [hunk 2]
```diff
@@ -18,6 +20,58 @@ def eager(gm, fake_tensor_inputs):
     return gm
 
 
+@register_backend(name="explain")
+def explain(gm: torch.fx.GraphModule, example_inputs):
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:24]: The explain backend needs to be composable with other backends. See CompileCounterWithBackend for an example.

File: torch/_dynamo/config.py [hunk 3]
```diff
@@ -108,6 +108,11 @@
 # Show a warning on every graph break
 print_graph_breaks = False
 
+# This flag will enable logging nodes while
+# the explain backend is traversing the module
+# for debugging-purposes
+explain_print_graphs = False
```
Comment (voznesenskym) [inline torch/_dynamo/config.py:114]: Not a general config, should be an arg to the explain backend.

Comment (alimoezzi) [reply torch/_dynamo/config.py:114]: Again, this is what I also prefer but that's the the signature of `compiler_fn` that I need to follow:

```
#_dynamo/backends/registry.py
CompilerFn = Callable[[fx.GraphModule, List[torch.Tensor]], CompiledFn]
```

File: torch/_dynamo/backends/debugging.py [hunk 4]
```diff
@@ -18,6 +20,58 @@ def eager(gm, fake_tensor_inputs):
...
+    """
+    if config.explain_print_graphs:
+        try:
+            from tabulate import tabulate
+        except ImportError:
+            print(
+                "`Explain Backend` relies on the library `tabulate`, "
+                "which could not be found on this machine. Run `pip "
+                "install tabulate` to install the library."
+            )
```
[the remark this answers is not in the document]
Comment (alimoezzi) [reply torch/_dynamo/backends/debugging.py:54]: Not sure what you mean by "Isn't this an error from print_tabular"?
Here I need to format the graph node so I need to import `tabular` for this functionality and raise error if the user has not `tabular` installed.

File: torch/_dynamo/backends/debugging.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alimoezzi) [reply torch/_dynamo/backends/debugging.py:70]: Again, I would also prefer this approach.
Yet, the goal of the original issue is to choose this compiler by string in a very convenient way.
```
opt_fn = torch.compile(fn, backend="explain")
```
and with the API that dynamo currently has for `compile`, `optimize` and `register_backend`, it was not possible without major API change.

[the remark this answers is not in the document]
Comment (alimoezzi) [reply torch/_dynamo/backends/debugging.py:70]: > Short term task
Introduce a new backend - explain such that torch.compile(fn, backend="explain") gives morally same output as torch._dynamo.explain. We can add the explain backend in this file - [torch/_dynamo/backends/debugging.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/backends/debugging.py)

> As a start, we can focus on graph breaks and explanation string. We would want the backend to be configurable so that we can expand to other stuff.

This is the short term plan for this issue. @voznesenskym

Comment (alimoezzi) [reply torch/_dynamo/backends/debugging.py]: Output is updated with break reason number.
```
...
Graph Break
Break reason 4:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 15, in fn3
    torch._dynamo.graph_break()
```

**Comment by alimoezzi** [comment]
@voznesenskym 
Thanks for the comments. Let me know to what extend, you want to keep this backend simple without requiring any change to other APIs.

**Comment by alimoezzi** [comment]
Thanks @msaroufim all tests are green.
cc @voznesenskym

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Done here https://github.com/pytorch/pytorch/pull/103259/files

**Comment by ezyang** [comment]
obsoleted by https://github.com/pytorch/pytorch/pull/103259
